### ISSUE TITLE
Unbound DNSBL - Add option to use local-zone with refuse instead of redirecting everything to localhost via local-data

### ISSUE THREAD

**Issue by defiantmonkey** [opnsense/core#4557]
**Important notices**
Before you add a new report, we ask you kindly to acknowledge the following:

[x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md

[x] I have searched the existing issues and I'm convinced that mine is new.

**Is your feature request related to a problem? Please describe.**
Hi, new to OPNsense (and BSD in general), so forgive me if I sound dumb (though I probably am dumb given I'm new to a lot of this).

I mainly got OPNsense installed for a replacement of my local DNS server (plan to do more with it in the future), in particular I wanted to make use of the blocking feature in Unbound.

After testing it out, I noticed that when the dnsbl.conf file is generated everything is redirected to localhost. At a base level this works fine for blocking requests, but I also want to be able to go into the logs and see what was blocked. Currently it will just show that the request succeeded.

**Describe the solution you'd like**

I achieved this locally by modifying line 140 of https://github.com/opnsense/core/blob/master/src/opnsense/scripts/unbound/download_blacklists.py on my install so that it now looks like this:

unbound_outf.write("local-zone: \"%s\" refuse\n" % entry)

which lets me simply search the logs for the word "refuse" to see what is actually being blocked.

As I am quite new at doing this kind of stuff, this may not even be a good way of doing this, but it would be nice if a change could be made so that tracking blocked requests through the logs/statistics is possible on future releases.

Edit: After further discussion, I realized that others may have existing configs working with local-data entries, so instead of a hard change, a toggle/option to use local-zone with refuse instead of local-data redirect (probably through the web UI) would be a better solution.

**Describe alternatives you've considered**
I don't know of any alternative solutions so far.

EDIT: as discussed below, a current workaround for those who wish to stick to the web UI or have existing configs working with the local-data entries is to use the suricata alerts (Intrusion Detection service) and enable/disable the alert as needed. 

**Additional context**
Version: OPNsense 20.7.7_1-amd64

**Comment by defiantmonkey** [comment]
Edited the initial comment, I'm guessing this was marked incomplete because I didn't follow that exact template?

**Comment by kulikov-a** [comment]
@defiantmonkey 
sorry, and what if the client have backup (second) dns server configured?

**Comment by defiantmonkey** [comment]
@kulikov-a 
I feel that's out of scope since that would be more a configuration issue on the client side than anything else. To explain a little more, if you're setting dns servers on a client and your backup dns servers don't block the same things as your primary dns server, that should be something that whoever is setting up the dns servers on a client should worry about. 

To me it doesn't make sense that a DNS server should have to try and compensate for other servers the client may or may not be using.

If this is a mistaken view please explain why as I am not that knowledgable in regards of DNS.

If it turns out that's something that OPNsense should be worrying about, then another solution I can propose is a toggle for the blocking mode (local-data redirect to localhost being default, toggle on to local-zone refuse). This might be a better solution actually so that administrators can configure it based on their own needs.

**Comment by kulikov-a** [comment]
@defiantmonkey 
I do not think this configuration type is wrong: it is normal practice to have a redundant server, even with less functionality. and most importantly, the proposed change will kill the already working dnsbl configurations with the described redundant servers configs.

>  This might be a better solution actually so that administrators can configure it based on their own needs.
I would have liked the option with a switch or some other way of customizing records too. but I'm not sure if the devs will like this)

>The end goal is to have some sort of way to track what's being blocked
i think about suricata rule for this. not tested yet. will update info

**Comment by kulikov-a** [comment]
might work. quickly checked by making a suricata rule, adding the github to the blacklist and ping:
![suricata_dnsbl](https://user-images.githubusercontent.com/36099472/103365620-8aaac600-4ad1-11eb-9e28-6ff27725e487.PNG)

**Comment by defiantmonkey** [comment]
Thanks for the explanation, makes more sense now, I cerainly didn't think about others who have already configured things around the current local-data when exploring this.

The alert idea is certainly adequate, and is a more persistent workaround than what I've done, but in my case would not be desirable as I don't want to be alerted every time something is blocked, this would trigger many alerts that I would rather just ignore (ie: during general web browsing).

Probably a little more context I could have added in the beginning, the reason I want the ability to tell if a request was actually blocked or went through is mostly in the case of connecting new devices to a network or installing new software so that I can see which connections I actually want/have to allow vs which connections I don't want/need.

I suppose in my particular case I can get away with configuring a cronjob through the terminal to update that line in the script every time the file is updated (or some other similar method), though having a more officialy supported method of configuration would be nice.

Out of curiosity, do you know why the devs wouldn't be open to the idea of allowing either a toggle between refuse/redirect or some more configurable setting for the DNSBL?

**Comment by kulikov-a** [comment]
> why the devs wouldn't be open to the idea of allowing either a toggle between refuse/redirect or some more configurable setting for the DNSBL?

I am not suggesting that they will not be open to idea. I just do not know.
For me, any changes should not kill existing configurations, and any complications in configuration should be seriously justified and verified.

>  I want the ability to tell if a request was actually blocked or went through is mostly in the case of connecting new devices to a network or installing new software so that I can see which connections I actually want/have to allow vs which connections I don't want/need

sorry, imho "suricata alert workaround " should still work, no? you can use (enable) this suricata alert on demand and just collate unbound dns request logs with suricata alerts to understand what was happening. and disable alert when not needed.

in any case - good luck with the FR. some way of tracking DNSBL hits would be very helpful.

**Comment by defiantmonkey** [comment]
> I am not suggesting that they will not be open to idea. I just do not know.

Ah okay, sorry for misunderstanding.

> For me, any changes should not kill existing configurations, and any complications in configuration should be seriously justified and verified.

I agree, I also believe that should be the case all the time

> sorry, imho "suricata alert workaround " should still work, no? you can use (enable) this suricata alert on demand and just collate unbound dns request logs with suricata alerts to understand what was happening. and disable alert when not needed.

Yeah that's definitely also a solution, although the only OPNsense service I'm using so far is Unbound (I plan to do more with OPNsense in the future, this was just a more immediate need for me), so I'd rather not have to configure other services just to have that one feature for one service. The sircata alert workaround definitely works for others who want to stick to the web interface though (or those who have other existing configs relying on local-data redirect).

I will change the request to be an option rather than a hard change.

**Comment by andmcg1** [comment]
I played around with all of the "type" options to find what would show the info needed for whitelisting false-positives within the Unbound logs. I found using this line works best and doesn't even require you to switch on verbose logging:

unbound_outf.write("local-zone: \"%s\" inform_redirect\n" % entry)
(File location: /usr/local/opnsense/scripts/unbound/download_blacklists.py)

You could even send this out to a syslog to see when a client may be trying to access malicious websites.

**Comment by kulikov-a** [comment]
@andmcg1 
interesting but its sets zone type to 'redirect' (plus inform) and now zone type for local-data is transparent.
'redirect; type will blacklist all subdomains afaik
so its possible to make it like
`
           unbound_outf.write("local-zone: \"%s\" inform\n" % entry)    
           unbound_outf.write("local-data: \"%s A 0.0.0.0\"\n" % entry)
`
and filter log by " inform " keyword. but it will make dnsbl.conf file larger

**Comment by andmcg1** [comment]
Ideally, if unbound supported more "inform" types then it could be done in a one-liner. The issue I noticed with the only other supported combination type "inform_deny" is the DNS query is not answered (ie, the client has to wait for a timeout), which could cause performance issues.

They great thing about "inform" is that it always logs it, even on default settings - you don't need to turn on verbose logging.

I tried your solution and it works well for my needs. The only small issue is that for example "dodgy.domain{dot}tv" is in the blacklist. If a client visits "domain{dot}tv" then Unbound logs this event as well. That's an Unbound issue, but just something to note.

I don't want to hijack this thread completely, but a simple solution I see is that a GUI option switches on and off that "inform" line. It could be named "Log blacklist hits" or similar

**Comment by kulikov-a** [comment]
@andmcg1 

> One caveat (but this may be an Unbound issue) is that "domain{dot}com" is on your blacklist, but "admin{dot}domain{dot}com" is not.
yes, this is the default behavior (the same as with the current settings) for cases of local-data without local-zone (in this case, a transparent local-zone is created. and 'inform' type is actualy transparent type plus inform). and in my opinion it is more correct since it excludes accidental blacklisting of subdomains.

> a simple solution I see is that a GUI option switches on and off that "inform" line.
I agree. it would be possible to add a "Inform zone type" checkbox with the corresponding help text

**Comment by AdSchellevis** [comment]
@kulikov-a I'm not sure what you're proposing, but if you open a PR I'll take a look

**Comment by kulikov-a** [comment]
@AdSchellevis 
thanks! will make PR and ticket at unbound for some shorthand of making local-data zone of type 'inform'
and I edited the picture highlighting the proposed addition

**Comment by pandaedward** [comment]
Hi team, sorry for keep coming back to the old issue. @kulikov-a #4715 is till pending for a merge but a suricata rule may work well for my infrastructure. Can you walk me through how to create an alert on specific port (53) please? My Intrsion Detection user defined rule has no port option. Im running `OPNsense 21.1.8_1-amd64`

**Comment by kulikov-a** [comment]
@pandaedward 
hi! 
I do not think it is possible to do this with GUI rules.
I think you need to add ruleset metadata to script folder
(like https://forum.opnsense.org/index.php?topic=7209.0)
and then create the file with rules on the server
for https://github.com/opnsense/core/issues/4557#issuecomment-752673121 i made a rule like:
`alert udp any 53 -> $HOME_NET any (msg:"DNS Reply with 0.0.0.0 (DNSBL record requested)"; content:"|00 01 00 01|"; content:"|00 04 00 00 00 00|"; distance:4; within:6; classtype:bad-unknown; sid:1000005; rev:3;)`
(https://github.com/kulikov-a/rules/blob/main/custom.bf.rules)

**Comment by pandaedward** [comment]
Thanks @kulikov-a that's really helpful, I put up a nginx config and opnsense picks up the custom ruleset straight away and the DNSBL alerts are showing beautifully. 

The problem is I don't see the payload in the alert so I can't identify which site was blacklisted. Am I missing anything in the configuration?

Again thanks heaps for the help.

**Comment by kulikov-a** [comment]
@pandaedward 
is "log package payload" enabled in Services: Intrusion Detection: Administration (with "advanced mode" on)?

**Comment by pandaedward** [comment]
That's it! All resolved. Suricata alert works well in my case and can be an excellent alternative to pihole. I am trying to cut down devices at the moment including those running as containers. Thanks so much!

**Comment by coach1988** [comment]
I just came across this while trying to figure out why my amazon app on android kept showing "something went wrong". I have  edited the "usr/local/opnsense/scripts/unbound/blocklists.py" (this will probably get overwritten again) to look like this (more escaping):
`unbound_outf.write("local-zone: \"%s\" refuse\n" % entry)`
Simply refusing instead of sending it to 0.0.0.0 (that's how i understood it) already seems to have solved this particular issue here.
On a side note: I had to set the log level to debug to show up.
